src/Components/WebAssembly/WebAssembly/src/Hosting/RootComponentTypeCache.cs

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Hosting
+{
+    internal static class RootComponentTypeCache
+    {
+        private static readonly Dictionary<Key, Type> _typeToKeyLookUp = new ();
+
+        public static Type GetRootComponent(string assembly, string type)
+        {
+            var key = new Key(assembly, type);
+            if (_typeToKeyLookUp.TryGetValue(key, out var resolvedType))
+            {
+                return resolvedType;
+            }
+            else
+            {
+                resolvedType = ResolveType(key, AppDomain.CurrentDomain.GetAssemblies());
```

Comment:
Can you add a comment explaining why it's guaranteed for the component assembly to have already been loaded?

Reply:
It is explicitly called out in the spec that lazy loaded pre-registered root components are not supported. I've unified this type cache with the one on the server, so I'll leave it out.

Reply:
Switched to a unified cache type